File: TYPESCRIPT/objetLouable.ts
class ObjetLouable {
    nom: string;
    reservations: Reservation[];

    constructor(nom: string) {
        this.nom = nom;
        this.reservations = [];
    }

    estDisponible(debut: string | Date, duree: number): boolean {
        const tentative = new Reservation(this, debut, duree);
        return this.reservations.every(res => !res.chevauche(tentative));
    }

    reserver(debut: string | Date, duree: number): boolean {
        if (duree < 1) {
            console.error(`Erreur : la durée minimale est de 1 jour (tenté avec ${duree}).`);
            return false;
        }

        if (!this.estDisponible(debut, duree)) {
            console.error(`Erreur : "${this.nom}" est déjà réservé à cette période.`);
            return false;
        }

        const nouvelleReservation = new Reservation(this, debut, duree);
        this.reservations.push(nouvelleReservation);
        console.log(`Réservation confirmée : "${this.nom}" du ${nouvelleReservation.debut.toDateString()} au ${nouvelleReservation.fin.toDateString()}`);
        return true;
    }

    afficherReservations(): void {
        console.log(`Réservations pour "${this.nom}" :`);
        if (this.reservations.length === 0) {
            console.log("  Aucune réservation.");
        } else {
            this.reservations.forEach((r, i) => {
                console.log(`  ${i + 1}. Du ${r.debut.toDateString()} au ${r.fin.toDateString()}`);
            });
        }
    }
}

File: JAVA/ObjetLouable.java
package JAVA;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class ObjetLouable {
    private String nom;
    private List<Reservation> reservations;

    public ObjetLouable(String nom) {
        this.nom = nom;
        this.reservations = new ArrayList<>();
    }

    public boolean estDisponible(LocalDate debut, int duree) {
        Reservation tentative = new Reservation(this, debut, duree);
        for (Reservation r : reservations) {
            if (r.chevauche(tentative)) {
                return false;
            }
        }
        return true;
    }

    public boolean reserver(LocalDate debut, int duree) {
        if (duree < 1) {
            System.out.println("Erreur : la durée minimale est de 1 jour.");
            return false;
        }

        if (!estDisponible(debut, duree)) {
            System.out.println("Erreur : \"" + nom + "\" est déjà réservé à cette période.");
            return false;
        }

        Reservation nouvelle = new Reservation(this, debut, duree);
        reservations.add(nouvelle);
        System.out.println("Réservation confirmée : \"" + nom + "\" du " + debut + " au " + nouvelle.getFin());
        return true;
    }

    public void afficherReservations() {
        System.out.println("Réservations pour \"" + nom + "\" :");
        if (reservations.isEmpty()) {
            System.out.println("  Aucune réservation.");
        } else {
            int i = 1;
            for (Reservation r : reservations) {
                System.out.println("  " + i + ". Du " + r.getDebut() + " au " + r.getFin());
                i++;
            }
        }
    }
}
